### ISSUE TITLE
channel & sockets don't have a good error reporting hook

### ISSUE THREAD

**Issue by iwarshak** [phoenixframework/phoenix#4330]
### Environment

* Elixir version (elixir -v): 1.11.2
* Phoenix version (mix deps): 1.5.9
* Node.js version (node -v): 14.15.3
* NPM version (npm -v): 6.14.9
* Operating system:

### Expected behavior

As I was debugging potential LiveView connection issues with customers, I started looking at a way to catch and report JS errors to honeybadger. I created a [sample app](https://github.com/iwarshak/phx-socket-error) to demonstrate how the channel & socket errors work, and here is my JS code for socket & channels.

```
socket.onOpen(() => console.log("socket onOpen"))
socket.onClose(() => console.log("socket onClose"))
socket.onError((e) => console.log("socket onError", e, this))
socket.connect()

let channel = socket.channel("main", {})
channel.onClose(() => console.log("channel onClose"))
channel.onError(() => console.log("channel onError"))
channel.join()
  .receive("ok", resp => { console.log("channel joined successfully", resp) })
  .receive("error", resp => { console.log("Unable to join", resp) })
```

If I run this app, and simply ctrl-c the app server to simulate a network error, here is what I see in the chrome console (server goes away at `14:34:04`)

<img width="888" alt="Screen Shot 2021-05-16 at 2 34 20 PM" src="https://user-images.githubusercontent.com/471/118410214-f3937480-b653-11eb-97a5-b08595233306.png">

**Here's some of the problems I see**

1) There's no real way for me to report this error to an error reporting service. In the past, I naively pushed a honeybadger event inside `socket.onError`, but this gets called repeatedly as the client tries to reconnect and I quickly overran my honeybadger limit. I thought about taking advantage of the `tries` counter in phoenix.js to only report the error on the first retry but it's not really accessible. Then I thought of reporting the error at the channel level (`channel.onError`) since that hook only get's called when the channel initially died, but that hook has zero context about the error.

So, tldr; there's not a way to handle an initial error, without also handling all of the reconnect errors. 

2) These are less important, in my opinion than the above issue, but the callbacks are not as 'symmetrical' as I'd expect:

 - When the connection dies, the `channel.onError` get's called (expected), but the `socket.onClose` is called. I'd expect both `.onErrors` to be called.  

- When a reconnect fails, both `socket.onError` and `socket.onClose` are called. I'd expect only the `onError`

For both of these issues I fully expect that there is something I am missing or unaware of, but I thought I'd bring this up because I haven't seen it referred to anywhere else.

**Comment by chrismccord** [comment]
The `onError` callback now receives the number of established connections, so you can use this to determine if the connection was previously established and drive errors off of this fact, for example:

```javascript
socket.onError((error, transport, establishedConnections) => {
  if(establishedConnections === 0){
    // report if  connection failed to be established
  }
}
```

We should still fire onError in the general case since users can key off of it for UI connection state updates. Thanks!
